File: app/src/main/java/com/sdex/activityrunner/ApplicationsListAdapter.java
package com.sdex.activityrunner;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.RequestOptions;
import com.sdex.activityrunner.db.ActivityModel;
import com.sdex.activityrunner.db.ApplicationModel;
import com.sdex.activityrunner.db.ItemModel;
import com.sdex.activityrunner.util.GlideApp;
import java.util.ArrayList;
import java.util.List;

public class ApplicationsListAdapter extends BaseExpandableListAdapter {

  private List<ItemModel> items;
  private Context context;
  private RequestManager glide;

  public ApplicationsListAdapter(Context context) {
    this.context = context;
    this.items = new ArrayList<>();
    this.glide = GlideApp.with(context);
  }

  @Override
  public Object getChild(int groupPosition, int childPosition) {
    return this.items.get(groupPosition).getActivityModels().get(childPosition);
  }

  @Override
  public long getChildId(int groupPosition, int childPosition) {
    return childPosition;
  }

  @Override
  public View getChildView(int groupPosition, int childPosition, boolean isLastChild,
    View convertView, ViewGroup parent) {
    ActivityModel activityModel = (ActivityModel) getChild(groupPosition, childPosition);
    LayoutInflater inflater = LayoutInflater.from(context);
    View view = inflater.inflate(R.layout.item_application, parent, false);

    TextView text1 = view.findViewById(android.R.id.text1);
    text1.setText(activityModel.getName());

    TextView text2 = view.findViewById(android.R.id.text2);
    text2.setText(activityModel.getComponentName().getShortClassName());

    ImageView icon = view.findViewById(android.R.id.icon);
    glide.load(activityModel.getIconPath())
      .apply(new RequestOptions()
        .fitCenter())
      .into(icon);

    return view;
  }

  @Override
  public int getChildrenCount(int groupPosition) {
    return this.items.get(groupPosition).getActivityModels().size();
  }

  @Override
  public Object getGroup(int groupPosition) {
    return this.items.get(groupPosition);
  }

  @Override
  public int getGroupCount() {
    return this.items.size();
  }

  @Override
  public long getGroupId(int groupPosition) {
    return groupPosition;
  }

  @Override
  public View getGroupView(int groupPosition, boolean isExpanded, View convertView,
    ViewGroup parent) {
    ItemModel itemModel = (ItemModel) getGroup(groupPosition);
    final ApplicationModel applicationModel = itemModel.getApplicationModel();
    LayoutInflater inflater = LayoutInflater.from(context);
    View view = inflater.inflate(R.layout.item_activity, parent, false);

    TextView text = view.findViewById(android.R.id.text1);
    text.setText(applicationModel.getName());

    TextView text2 = view.findViewById(android.R.id.text2);
    text2.setText(applicationModel.getPackageName());

    ImageView icon = view.findViewById(android.R.id.icon);
    glide.load(applicationModel.getIconPath())
      .apply(new RequestOptions()
        .fitCenter())
      .into(icon);

    return view;
  }

  @Override
  public boolean hasStableIds() {
    return false;
  }

  @Override
  public boolean isChildSelectable(int groupPosition, int childPosition) {
    return true;
  }

  public void addItems(List<ItemModel> itemModels) {
    this.items.clear();
    this.items.addAll(itemModels);
    notifyDataSetChanged();
  }
}

File: app/src/main/java/com/sdex/activityrunner/app/ApplicationsListAdapter.kt
package com.sdex.activityrunner.app

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isInvisible
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions
import com.sdex.activityrunner.databinding.ItemApplicationBinding
import com.sdex.activityrunner.db.cache.ApplicationModel
import com.sdex.activityrunner.glide.GlideApp
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView

class ApplicationsListAdapter(
    activity: FragmentActivity
) : ListAdapter<ApplicationModel, ApplicationsListAdapter.AppViewHolder>(DIFF_CALLBACK),
    FastScrollRecyclerView.SectionedAdapter {

    private val glide = GlideApp.with(activity)

    var isShowSystemAppIndicator: Boolean = false
        @SuppressLint("NotifyDataSetChanged")
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var itemClickListener: ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return AppViewHolder(ItemApplicationBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
        holder.bind(getItem(position), glide, isShowSystemAppIndicator, itemClickListener)
    }

    override fun getSectionName(position: Int): String {
        val name = getItem(position).name
        return if (name.isNullOrEmpty()) "" else name.first().uppercaseChar().toString()
    }

    interface ItemClickListener {

        fun onItemClick(item: ApplicationModel)

        fun onItemLongClick(item: ApplicationModel)
    }

    class AppViewHolder(
        private val binding: ItemApplicationBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(
            item: ApplicationModel,
            glide: RequestManager,
            isShowSystemAppIndicator: Boolean,
            itemClickListener: ItemClickListener?,
        ) {
            binding.name.text = item.name
            binding.packageName.text = item.packageName
            binding.system.isInvisible = !(item.system && isShowSystemAppIndicator)

            glide.load(item)
                .apply(RequestOptions().fitCenter())
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(binding.icon)

            binding.root.setOnClickListener {
                itemClickListener?.onItemClick(item)
            }
            binding.root.setOnLongClickListener {
                itemClickListener?.onItemLongClick(item)
                true
            }
            binding.appMenu.setOnClickListener {
                itemClickListener?.onItemLongClick(item)
            }
        }
    }

    companion object {

        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ApplicationModel>() {

            override fun areItemsTheSame(
                oldItem: ApplicationModel,
                newItem: ApplicationModel
            ): Boolean {
                return oldItem.packageName == newItem.packageName
            }

            override fun areContentsTheSame(
                oldItem: ApplicationModel,
                newItem: ApplicationModel
            ): Boolean {
                return oldItem == newItem
            }
        }
    }
}
